apis/telemetry/v1alpha1/telemetry_methods.go

```diff
@@ -0,0 +1,99 @@
+package v1alpha1
+
+import (
+	"errors"
+
+	"reflect"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/go-logr/logr"
+	promapi "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+// Merge takes an object a and merges another object, b's values into it, overwriting any which conflict.
+func (a *TelemetrySpec) Merge(b *TelemetrySpec) *TelemetrySpec {
+	// TODO: This method is brittle. It must be updated whenever any field is added to `TelemetrySpec``. It
+	// would be best to replace these methods with a generic client side `Merge()` based on the k8s strategic merge
+	// patch logic.
+	out := TelemetrySpec{}
+	if a == nil && b == nil {
+		return &out
+	} else if b == nil {
+		return a
+	} else if a == nil {
+		return b
+	}
+	if b.Prometheus == nil {
+		out.Prometheus = a.Prometheus
+	} else {
+		out.Prometheus = a.Prometheus.Merge(b.Prometheus)
+	}
+	return &out
+}
+
+// Merge takes an object a and merges another object, b's values into it, overwriting any which conflict.
+func (a *PrometheusTelemetrySpec) Merge(b *PrometheusTelemetrySpec) *PrometheusTelemetrySpec {
+	// TODO: This method is brittle. It must be updated whenever any field is added to `PrometheusTelemetrySpec`. It
+	// would be best to replace these methods with a generic client side `Merge()` based on the k8s strategic merge
+	// patch logic.
+	out := PrometheusTelemetrySpec{}
+	if a == nil && b == nil {
+		return nil
+	} else if b == nil {
+		return a
+	} else if a == nil {
+		return b
+	}
+	if a.CommonLabels != nil {
+		for k, v := range a.CommonLabels {
+			out.CommonLabels[k] = v
+		}
+	}
+	if b.CommonLabels != nil {
+		for k, v := range b.CommonLabels {
+			out.CommonLabels[k] = v
+		}
+	}
+	if b.Enabled == nil {
+		out.Enabled = a.Enabled
+	}
+	return &out
+}
+
+func (tspec *TelemetrySpec) IsValid(client client.Client, logger logr.Logger) (bool, error) {
```

Comment:
I'm not sure it's a good idea to import `client.Client` into an api package.

Reply:
Can you expand on why that might be harmful?

Reply:
OK, I'll move this elsewhere. It seemed nice to be able to write `MyTelemetrySpec.IsValid()` but having to write `TelemetrySpecIsValid(MyTelemetrySpec)` is not the end of the world...